src/Algebra/Graph/Labelled/AdjacencyMap.hs

```diff
@@ -237,16 +232,20 @@ infixl 5 >-
 -- 'edgeLabel' x z $ 'transitiveClosure' (overlay ('edge' e x y) ('edge' 'one' y z)) == e
 -- 'edgeLabel' x z $ 'transitiveClosure' (overlay ('edge' e x y) ('edge' f   y z)) == e '<.>' f
 -- @
-overlay :: (Eq e, Monoid e, Ord a) => AdjacencyMap e a -> AdjacencyMap e a -> AdjacencyMap e a
-overlay (AM x) (AM y) = AM $ Map.unionWith nonZeroUnion x y
+overlay :: (Semigroup e, Ord a) => AdjacencyMap e a -> AdjacencyMap e a -> AdjacencyMap e a
+overlay (AM x) (AM y) = AM $ Map.unionWith (Map.unionWith (<>)) x y
 
--- Union maps, removing zero elements from the result.
-nonZeroUnion :: (Eq e, Monoid e, Ord a) => Map a e -> Map a e -> Map a e
-nonZeroUnion x y = Map.filter (/= zero) $ Map.unionWith mappend x y
-
--- Drop all edges with zero labels.
-trimZeroes :: (Eq e, Monoid e) => Map a (Map a e) -> Map a (Map a e)
-trimZeroes = Map.map (Map.filter (/= zero))
+-- | An 'AdjacencyMap' represents the absense of an edge by excluding the edge
+-- from the underlying set of adjacencies.  This is both for performance reasons
+-- and also to allow edge types that are 'Semigroup' but /not/ 'Monoid' (that
+-- is, they have no 'zero' value).  For edge types that do have a 'zero' value,
+-- then for performance reasons, it is optimal to exclude all 'zero' edges.
+--
+-- 'trimZeros x' is behaviorally an identity function, but it should be used for
+-- performance reasons when a monoidal edge type is used and 'zero' edges may
+-- have slipped into the underlying representation.
+trimZeroes :: (Eq e, Monoid e) => AdjacencyMap e a -> AdjacencyMap e a
+trimZeroes (AM x) = AM $ Map.map (Map.filter (/= zero)) x
```

Comment:
Exposing this seems like an abstraction leak.  However, keeping it hidden could cause users to suffer from performance penalties if their edge type _has_ `zero`.

Reply:
Given that the implementation doesn't care much about zeroes, should we provide a more general function like `filterEdges`?

Reply:
I want to propose a radical alternative.  Instead of exposing `trimZeros` (or `filterEdges`), we make the monoid explicit and this function implicit.  I feel like Haskell's type classes are not quite equipped to deal with this, but I can think of a work around.  We could define:
```haskell
data AdjacencyMap e a = AM {
    -- | The /adjacency map/ of an edge-labelled graph: each vertex is
    -- associated with a map from its direct successors to the corresponding
    -- edge labels.
    adjacencyMap :: Map a (Map a e),
    zeroEdge :: Maybe e
  } deriving (Eq, Generic, NFData)
```
Then, we could have two constructors, one that takes a Semigroup constraint and sets `zeroEdge = Nothing` and the other that takes a Monoid constraint and sets it to `mempty`.  We'd add back in all the uses of `filterZeros`, but they'd be guarded by whether `zeroEdge` was `Nothing` or not.